### DISCUSSION TITLE
Custom vim keybinding to quickly delete content in quotes, brackets, etc.

### DISCUSSION THREAD

**Discussion by speri203** [zed-industries/zed#6658] [Feature Requests]
Recently came across this editor and I have been impressed by the speed of it. I generally use neovim as my main editor 99% and if I am able to incorporate my frequently used keybindings then I can attempt to daily drive Zed. The keybindings I would like to incorporate are:

jj - to substitute for Esc (use a delay to determine if actually typing jj or to use Escape)
ciq/diq - to clear and insert or delete content inside single/double quotes (same for yanking)
cib/dib - to clear and insert or delete content within brackets/parentheses/curly braces (same for yanking)

I have come across the documentation on the website outlining how to create custom vim keybindings but I am not able to find any built in API "vim::" call to accomplish this. Has anyone been able to figure this out/help me out?

**Comment by danielkvores** [comment]
I have the same issue with jj and I can't understand why it doesn't work. I've tried alternatives like jk, but still not luck, when I'm in insert mode the j won't be displayed which is weird because g works perfectly fine in insert mode even though there are multiple commands utilising g, f.e. gn

**Reply by mrnugget** [reply]
Just FYI: this is currently being fixed. In an upcoming Zed release you should then be able to bind `jk` to switch to normal mode, for example.

**Comment by speri203** [comment]
> Just FYI: this is currently being fixed. In an upcoming Zed release you should then be able to bind `jk` to switch to normal mode, for example.

Awesome, looking forward to the update. Do you happen to have any updates regarding the 2nd and 3rd keymaps I have mentioned in the post? Chain actions for yanking/clearing/deleting content found within special characters? The functionality already semi exists. For example, if I want to change content within single quotes then I would do `ci'` and for double quotes `ci"`. What I would like to do is `ciq` which would "change" content within quotes irrespective of single or double. Similar behavior for `cib` which would change content in `(),[],{}`

**Reply by mrnugget** [reply]
I don't think that's possible yet without changing the code. See the default [`vim.json`](https://zed.dev/ref/vim.json) here:

```json
{
    "context": "Editor && VimObject",
    "bindings": {
      "w": "vim::Word",
      "shift-w": [
        "vim::Word",
        {
          "ignorePunctuation": true
        }
      ],
      "s": "vim::Sentence",
      "'": "vim::Quotes",
      "`": "vim::BackQuotes",
      "\"": "vim::DoubleQuotes",
      "(": "vim::Parentheses",
      ")": "vim::Parentheses",
      "[": "vim::SquareBrackets",
      "]": "vim::SquareBrackets",
      "{": "vim::CurlyBrackets",
      "}": "vim::CurlyBrackets",
      "<": "vim::AngleBrackets",
      ">": "vim::AngleBrackets"
    }
  }
```

What we'd need is a `vim::AnyParenBracket` or `vim::AnyQuote` object. Doesn't seem to hard, but it doesn't work yet. Noted 😄

**Reply by speri203** [reply]
Awesome, thanks for the engagement! The `nvim-surround` plugin in neovim is what gives the capability of the `ciq, diq, cib, etc..` I figure id post the link here as well so you can have a reference https://github.com/kylechui/nvim-surround

**Comment by ConradIrwin** [comment]
Currently we're prioritizing vim compatibility over custom keybindings; that said, there are a number of commonly used vim plugins I'd like to emulate.

Is there a plugin that you use to provide the objects that you use? (I don't think `q` is defined in vim, so we could probably make it work as you suggest, `b` in vim and zed just matches `(` and `)`).

**Reply by speri203** [reply]
I use the pre-build LazyVim distribution and the functionality came out of the box but after some digging I believe I found the plugin responsible for this functionality, `nvim-surround`. The video demoing the plugin uses something similar `dsq` to remove quotes that wrap a word. 
https://github.com/kylechui/nvim-surround

**Comment by tusharhqq** [comment]
how to enable treesitter context in zed?
![image](https://github.com/zed-industries/zed/assets/79051850/88c80af3-6a0d-4458-b8f5-72838651627d)

**Reply by mrnugget** [reply]
Can you clarify what you're looking for? A text motion to select tree-sitter contexts? These already exist, you just need to copy them to your keybindings: https://github.com/zed-industries/zed/blob/ac74a72a9ef80a6cc50e7b001e163e76fbd9fa3d/assets/keymaps/default.json#L336-L337

**Comment by tusharhqq** [comment]
Got it
